docs: add setup anoncreds and indy-vdr (#107)
animo/aries-javascript-docs: guides/getting-started/set-up/indy-vdr.md

We recommend new projects to use Indy VDR from the start, and also to migrate
existing projects to Indy VDR, from the indy-sdk. However, projects may
experience some issues. If you encounter any issues, please [open an
issue](https://github.com/hyperledger/aries-framework-javascript/issues/new).

Currently, there are few limitations to using Indy VDR.

- React Native projects leveraging [Expo](https://expo.dev) are not supported
  at the moment. Support for Expo will be added soon.
- React Native Android support starts from 11 right now. Work is being done on
  support for lower versions.
- When running in Node.JS, only Node.JS 18 is suported for now. See [Supported
  Node.JS versions for Indy VDR](#supported-nodejs-versions-for-indy-vdr)

:::

### Installing Indy VDR

When using Aries Framework JavaScript with Indy VDR, there are a few extra
dependencies that need to be installed. We need to install the
`@hyperledger/indy-vdr`, which contains all the functionality to register
objects on an Hyperledger Indy VDR. Secondly, we need to add native bindings
for the specific platform `@hyperledger/indy-vdr-<platform>`. Currently there

> are bindings for Node.JS, `nodejs` and React Native, `react-native`.

Reviewer: same suggestion as on anoncreds doc

<!--tabs-->

# Node.JS

```console
yarn add @aries-framework/indy-vdr @hyperledger/indy-vdr-nodejs
```

# React Native

```console
yarn add @aries-framework/indy-vdr @hyperledger/indy-vdr-react-native
```

<!--/tabs-->

### Adding Indy VDR to the agent

After installing the dependencies, we can register the Indy VDR module on the
agent.

<!--tabs-->

# Node.JS

```typescript showLineNumbers set-up-indy-vdr.ts section-1

```

# React Native

```typescript showLineNumbers set-up-indy-vdr-rn.ts section-1

```

<!--/tabs-->

As you can see, the Indy VDR module takes the native bindings and a list of
networks. This list of networks will be used to resolve and register objects
on. The list of networks is compatible with the older

> [`indyLedgers`](../../tutorials/agent-config/index.md#indyLedgers).

Reviewer: I have a local PR that removes this from the docs, as the property is removed from the agent config in 0.4.0

Reviewer 2: I added a configuration section to the document. Its mainly copied from the `indyLedgers` from the old docs. I am not too sure on the structure, but if it looks fine to you, I will also add it to aries-askar and anoncreds-rs.

### Supported Node.JS versions for Indy VDR

Due to an issue in `ref-napi` (which is used in the Node.JS bindings for
Indy VDR), performance for Indy VDR in Node.JS is not as expected. A
patched version for `ref-napi` has been published that fixes this issue, but
this only works in Node.JS 18+.

To use Indy VDR in Node.JS, make sure you're using Node.JS 18 and patch the
`ref-napi` package to point towards `@2060.io/ref-napi` in your `package.json`:

<!--tabs-->

# NPM

> Using [NPM
> `overrides`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides)

Reviewer: ```suggestion
Using [NPM `overrides`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides)
```

we can point the `ref-napi` package to `@2060.io/ref-napi`.

```json
{
    "overrides" {
        "ref-napi": "npm:@2060.io/ref-napi"
    }
}
```

# Yarn

> Using [Yarn
> `resolutions`](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/)

Reviewer: ```suggestion
Using [Yarn `resolutions`](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/)
```

Currently, there are few limitations to using Indy VDR.

- React Native projects leveraging [Expo](https://expo.dev) are not supported at the moment. Support for Expo will be added soon.
- React Native Android support starts from 11 right now. Work is being done on support for lower versions.
- When running in Node.JS, only Node.JS 18 is suported for now. See [Supported Node.JS versions for Indy VDR](#supported-nodejs-versions-for-indy-vdr)

:::

### Installing Indy VDR

When using Aries Framework JavaScript with Indy VDR, there are a few extra dependencies that need to be installed. We need to install the `@hyperledger/indy-vdr`, which contains all the functionality to register objects on an Hyperledger Indy VDR. Secondly, we need to add native bindings for the specific platform `@hyperledger/indy-vdr-<platform>`. Currently there are bindings for Node.JS, as `@hyperledger/indy-vdr-nodejs`, and React Native as `@hyperlegder/indy-vdr-react-native`.

<!--tabs-->

# Node.JS

```console
yarn add @aries-framework/indy-vdr @hyperledger/indy-vdr-nodejs
```

# React Native

```console
yarn add @aries-framework/indy-vdr @hyperledger/indy-vdr-react-native
```

<!--/tabs-->

> ### Configuration

Reviewer: I think we should swap adding it to the agent and config (as you need to add it first before you can think about how to configure it)

Reviewer 2: I am fine with either. My idea was you first need to know how to configure it before you add it. Will swap.

Reviewer: also makes sense. Feel free to keep it as is

As you can see below, the Indy VDR module takes the native bindings and a list of networks. This list of networks will be used to resolve and register objects on.

```typescript title="example"
import { indyVdr } from '@hyperledger/indy-vdr-nodejs'
import { IndyVdrModule } from '@aries-framework/indy-vdr'

> indyVdr: new IndyVdrModule({

Reviewer: this feels a bit weird. Maybe we can do `const modules = ` and having the indyVdr module as the only key

  indyVdr,
  networks: [
    {
      indyNamespace: 'bcovrin:test',
      isProduction: false,

>       genesisPath: './path/to/genesis.txn',

Reviewer: I think indy vdr doesn't support `genesisPath` anymore? Also, should we use a snippet for this so we have type checking?

      connectOnStartup: true,
    },
  ],
})
```

#### indyVdr

**Type**: `IndyVdr`

> the `indyVdr` key takes a class that implements all the native bindings for Indy VDR. This can be extracted from the `@hyperledger/indy-vdr-nodejs` package or the `@hyperledger/indy-vdr-react-native` package.

Reviewer: ```suggestion
the `indyVdr` key takes a class that implements all the native bindings for Indy VDR. This can be imported from the `@hyperledger/indy-vdr-nodejs` package or the `@hyperledger/indy-vdr-react-native` package.
```

#### networks

**Type**: `IndyVdrPoolConfig[]`

> An array of indy networks to connect to. The list can contain the following object and it must include _either_ the [`genesisPath`](.md#indyledgersgenesispath) _or_ [`genesisTransactions`](#indyledgersgenesistransactions). It is important to know that the first ledger in the list ledgers will be used for registering the schema, credential definition, etc.

Reviewer: This is not true anymore, you now always need to provide the issuerId

```suggestion
An array of indy networks to connect to. The list can contain the following object and it must include _either_ the [`genesisPath`](.md#indyledgersgenesispath) _or_ [`genesisTransactions`](#indyledgersgenesistransactions).
```

##### indyNamespace

**Type**: `string`

The Indy namespace aka the name identifying the name of the network connecting to. See also [indy did method identifiers](https://hyperledger.github.io/indy-did-method/#indy-did-method-identifiers)

##### `isProduction`

**Type**: `boolean`

> Whether the ledger is a production ledger. This is used for the pick-up algorithm as production ledgers have priority.

Reviewer: ```suggestion
Whether the ledger is a production ledger. This is used for detecting which ledger to use in case of unqualified identifiers as production ledgers have priority.
```

> ##### `genesisPath`

Reviewer: as mentioned, this doesn't exist anymore

**Type**: `string`

Filesystem path of the genesis transaction. At this location, there will just be a JSON object like the [`genesisTransaction`](#genesistransactions).

##### `genesisTransactions`

**Type**: `string`

> Stringified JSON object of the transaction.

Reviewer: Nit, and we don't have to change it, but it's not really a json object, as it's a string with multiple json objects on new lines

##### `transactionAuthorAgreement`

**Type**: `TransactionAuthorAgreement`

JSON representation specifying the version and acceptance mechanism. The version is the unique version of the transaction author agreement acceptance mechanism list (AML). The acceptance mechanism refers to the acceptance mechanism label of the item in the AML. For more details you may consult the [indy-node docs on AML](https://github.com/hyperledger/indy-node/blob/master/docs/source/transactions.md#transaction_author_agreement_aml)

##### `transactionAuthorAgreement.version`

**Type**: `string`

The version of the AML acceptance mechanism. This is a string representation of a version number e.g. '1' or '1.4'

> ##### `indyLedgers.transactionAuthorAgreement.acceptanceMechanism`

Reviewer: The prefix levels are different here?
